File: typegraph/deno/sdk/src/runtimes/random.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import * as t from "../types.ts";
import { runtimes } from "../wit.ts";
import { RandomRuntimeData } from "../gen/typegraph_core.d.ts";
import { Materializer, Runtime } from "./mod.ts";
import { fx } from "../index.ts";

interface RandomMat extends Materializer {
  runtime: number;
}

export class RandomRuntime extends Runtime {
  constructor(data: RandomRuntimeData) {
    super(runtimes.registerRandomRuntime(data));
  }

  gen(out: t.Typedef) {
    const effect = fx.read();

    const matId = runtimes.createRandomMat(
      {
        runtime: this._id,
        effect,
      },
      {
        runtime: this._id,
      },
    );

    return t.func(t.struct({}), out, {
      _id: matId,
      runtime: this._id,
    } as RandomMat);
  }
}

File: src/typegraph/core-ng/src/runtimes/random.rs
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::{
    errors::Result,
    global_store::Store,
    types::{
        core::MaterializerId,
        runtimes::{BaseMaterializer, MaterializerRandom, RandomRuntimeData},
    },
};

use super::{Materializer, Runtime};

#[derive(Debug)]
pub enum RandomMaterializer {
    Runtime(MaterializerRandom),
}

pub fn register_random_runtime(data: RandomRuntimeData) -> Result<MaterializerId> {
    Ok(Store::register_runtime(Runtime::Random(data.into())))
}

pub fn create_random_mat(
    base: BaseMaterializer,
    data: MaterializerRandom,
) -> Result<MaterializerId> {
    let mat = Materializer::random(base.runtime, RandomMaterializer::Runtime(data), base.effect);
    Ok(Store::register_materializer(mat))
}
